File: src/user_examples/cs_user_boundary_conditions-richards.f90
!-------------------------------------------------------------------------------

!VERS

! This file is part of Code_Saturne, a general-purpose CFD tool.
!
! Copyright (C) 1998-2019 EDF S.A.
!
! This program is free software; you can redistribute it and/or modify it under
! the terms of the GNU General Public License as published by the Free Software
! Foundation; either version 2 of the License, or (at your option) any later
! version.
!
! This program is distributed in the hope that it will be useful, but WITHOUT
! ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
! FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
! details.
!
! You should have received a copy of the GNU General Public License along with
! this program; if not, write to the Free Software Foundation, Inc., 51 Franklin
! Street, Fifth Floor, Boston, MA 02110-1301, USA.

!-------------------------------------------------------------------------------

!===============================================================================
! Function:
! ---------
!> \file  cs_user_boundary_conditions-richards.f90
!> \brief Example of cs_f_user_boundary_conditions subroutine.f90 for
!>        groundwater flow module
!
!-------------------------------------------------------------------------------

!-------------------------------------------------------------------------------
! Arguments
!______________________________________________________________________________.
!  mode           name          role                                           !
!______________________________________________________________________________!
!> \param[in]     nvar          total number of variables
!> \param[in]     nscal         total number of scalars
!> \param[out]    icodcl        boundary condition code:
!>                               - 1 Dirichlet
!>                               - 2 Radiative outlet
!>                               - 3 Neumann
!>                               - 4 sliding and
!>                                 \f$ \vect{u} \cdot \vect{n} = 0 \f$
!>                               - 5 smooth wall and
!>                                 \f$ \vect{u} \cdot \vect{n} = 0 \f$
!>                               - 6 rough wall and
!>                                 \f$ \vect{u} \cdot \vect{n} = 0 \f$
!>                               - 9 free inlet/outlet
!>                                 (input mass flux blocked to 0)
!> \param[in]     itrifb        indirection for boundary faces ordering
!> \param[in,out] itypfb        boundary face types
!> \param[out]    izfppp        boundary face zone number
!> \param[in]     dt            time step (per cell)
!> \param[in,out] rcodcl        boundary condition values:
!>                               - rcodcl(1) value of the dirichlet
!>                               - rcodcl(2) value of the exterior exchange
!>                                 coefficient (infinite if no exchange)
!>                               - rcodcl(3) value flux density
!>                                 (negative if gain) in w/m2 or roughness
!>                                 in m if icodcl=6
!>                                 -# for the velocity \f$ (\mu+\mu_T)
!>                                    \gradt \, \vect{u} \cdot \vect{n}  \f$
!>                                 -# for the pressure \f$ \Delta t
!>                                    \grad P \cdot \vect{n}  \f$
!>                                 -# for a scalar \f$ cp \left( K +
!>                                     \dfrac{K_T}{\sigma_T} \right)
!>                                     \grad T \cdot \vect{n} \f$
!_______________________________________________________________________________

subroutine cs_f_user_boundary_conditions &
 ( nvar   , nscal  ,                                              &
   icodcl , itrifb , itypfb , izfppp ,                            &
   dt     ,                                                       &
   rcodcl )

!===============================================================================

!===============================================================================
! Module files
!===============================================================================

use paramx
use numvar
use optcal
use cstphy
use cstnum
use entsor
use parall
use period
use ihmpre
use ppppar
use ppthch
use coincl
use cpincl
use ppincl
use ppcpfu
use atincl
use ctincl
use cs_fuel_incl
use mesh
use field

!===============================================================================

implicit none

! Arguments

integer          nvar   , nscal

integer          icodcl(nfabor,nvar)
integer          itrifb(nfabor), itypfb(nfabor)
integer          izfppp(nfabor)

double precision dt(ncelet)
double precision rcodcl(nfabor,nvar,3)

! Local variables

!< [loc_var_dec]
integer          ifac, ilelt, nlelt, ii, iflmab
double precision Vel, Cref, R_act, S_act, T_act

integer, allocatable, dimension(:) :: lstelt
double precision, dimension(:), pointer :: bmasfl
!< [loc_var_dec]

!===============================================================================

!===============================================================================
! Initialization
!===============================================================================

allocate(lstelt(nfabor))  ! temporary array for boundary faces selection

!===============================================================================
! Assign boundary conditions to boundary faces here

! For each subset:
! - use selection criteria on group (face name) to filter boundary faces of a
!   given subset
! - loop on faces from a subset
! - set the boundary condition for each face
!===============================================================================

! For groundwater flow module, the undefined type face iindef is always used.
! BC for the flow part are on the pressure head H=h+z

! Example 1: Dirichlet on hydraulic head and solutes

!< [richards_BC_ex1]
call getfbr('FACE1', nlelt, lstelt)

do ilelt = 1, nlelt

! Face number
  ifac = lstelt(ilelt)

! Undefined type face
  itypfb(ifac) = iindef

! Velocity is not solved but deduced from darcy law, then no BC are required.
! However, no flux BC are imposed for safety.
  icodcl(ifac,iu) = 3
  icodcl(ifac,iv) = 3
  icodcl(ifac,iw) = 3
  rcodcl(ifac,iu,3) = 0.d0
  rcodcl(ifac,iv,3) = 0.d0
  rcodcl(ifac,iw,3) = 0.d0

! Dirichlet BC on hydraulic head (H = h + z) to impose a constant value
  icodcl(ifac,ipr) = 1
  rcodcl(ifac,ipr,1) = 10.d0

! Dirichlet BC on centration C (g/m^3)
  do ii = 1, nscal
    icodcl(ifac,isca(ii)) = 1
    rcodcl(ifac,isca(ii),1) = 1.d0
  enddo
enddo
!< [richards_BC_ex1]

!< [richards_BC_ex2]
call getfbr('FACE2', nlelt, lstelt)

do ilelt = 1, nlelt

! Face number
  ifac = lstelt(ilelt)

! Undefined type face
  itypfb(ifac) = iindef

! Velocity is not solved but deduced from darcy law, then no BC are required.
! However, no flux BC are imposed for safety.
  icodcl(ifac,iu) = 3
  icodcl(ifac,iv) = 3
  icodcl(ifac,iw) = 3
  rcodcl(ifac,iu,3) = 0.d0
  rcodcl(ifac,iv,3) = 0.d0
  rcodcl(ifac,iw,3) = 0.d0

! Dirichlet BC on hydraulic head (H = h + z) to impose a constant gradient over z axis
! here \f$ \grad H \cdot \vect{z} = -0.1 \f$
  icodcl(ifac,ipr) = 1
  rcodcl(ifac,ipr,1) = -1.d0 * cdgfbo(3,ifac)
enddo
!< [richards_BC_ex2]

! ---- Neumann on hydraulic head and solutes

!< [richards_BC_ex3]
call getfbr('FACE3', nlelt, lstelt)

do ilelt = 1, nlelt

! Face number
  ifac = lstelt(ilelt)

! Undefined type face
  itypfb(ifac) = iindef

! Velocity is not solved but deduced from darcy law, then no BC are required.
! However, no flux BC are imposed for safety.
  icodcl(ifac,iu) = 3
  icodcl(ifac,iv) = 3
  icodcl(ifac,iw) = 3
  rcodcl(ifac,iu,3) = 0.d0
  rcodcl(ifac,iv,3) = 0.d0
  rcodcl(ifac,iw,3) = 0.d0

! Neumann BC on hydraulic head (H = h + z) to impose a gradient among the surface normal
! Here \f$ \grad H \cdot \vect{n} = 0 \f$
  icodcl(ifac,ipr) = 3
  rcodcl(ifac,ipr,3) = 0.d0

! Neumann BC on centration C for boundary surface with outward or null normal flow $V_out$:
! It allows to impose a gradient among the surface normal for a diffusive flux as the
! convective flux is defined by the $C*V_out$.
! Here \f$ \grad C \cdot \vect{n} = 0 \f$
  do ii = 1, nscal
    icodcl(ifac,isca(ii)) = 3
    rcodcl(ifac,isca(ii),3) = 0.d0
  enddo
enddo
!< [richards_BC_ex3]

! ---- Mixed (or Robin) BC to impose a total flux (diffusive and convective flux)

!< [richards_BC_ex4]
call getfbr('FACE4', nlelt, lstelt)

! Pointers to the mass flux
call field_get_key_int(ivarfl(ipr), kbmasf, iflmab)
call field_get_val_s(iflmab, bmasfl)

do ilelt = 1, nlelt

! Face number
  ifac = lstelt(ilelt)

! Undefined type face
  itypfb(ifac) = iindef

! Velocity is not solved but deduced from darcy law, then no BC are required.
! However, no flux BC are imposed for safety.
  icodcl(ifac,iu) = 3
  icodcl(ifac,iv) = 3
  icodcl(ifac,iw) = 3
  rcodcl(ifac,iu,3) = 0.d0
  rcodcl(ifac,iv,3) = 0.d0
  rcodcl(ifac,iw,3) = 0.d0

! Dirichlet BC on hydraulic head (H = h + z) to impose a constant value
  icodcl(ifac,ipr) = 1
  rcodcl(ifac,ipr,1) = 10.d0

! In order to impose a radioactive activity R_act (Bq) at a surface S_act (m^2)
! during a period of time T_act, a mixed (or Robin) BC is used.
! To do so, two quantities are required:
! * the velocity at an entrance is Vel = - mass_flux / surf
!   or Vel = mass_flux / surf at an exit
! * the reference concentration Cref = R_act / (S_act * dt * Vel)
  R_act = 0.5d0
  S_act = 50.d0
  T_act = 10.d0

! The total flux is imposed from 0 to T_act
  Vel = - bmasfl(ifac)/surfbn(ifac)
  Cref = R_act / (S_act * dt(ntcabs) * Vel)
  if (ttcabs.gt.T_act) Cref = 0.d0
  do ii = 1, nscal
    icodcl(ifac,isca(ii)) = 1
    rcodcl(ifac,isca(ii),1) = Cref
    rcodcl(ifac,isca(ii),2) = Vel
  enddo
enddo
!< [richards_BC_ex4]

deallocate(lstelt)  ! temporary array for boundary faces selection

return
end subroutine cs_f_user_boundary_conditions

File: src/user_examples/cs_user_boundary_conditions-richards.c
/*============================================================================
 * User definition of boundary conditions.
 *============================================================================*/

/* VERS */

/*
  This file is part of code_saturne, a general-purpose CFD tool.

  Copyright (C) 1998-2022 EDF S.A.

  This program is free software; you can redistribute it and/or modify it under
  the terms of the GNU General Public License as published by the Free Software
  Foundation; either version 2 of the License, or (at your option) any later
  version.

  This program is distributed in the hope that it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
  details.

  You should have received a copy of the GNU General Public License along with
  this program; if not, write to the Free Software Foundation, Inc., 51 Franklin
  Street, Fifth Floor, Boston, MA 02110-1301, USA.
*/

/*----------------------------------------------------------------------------*/

#include "cs_defs.h"

/*----------------------------------------------------------------------------
 * Standard C library headers
 *----------------------------------------------------------------------------*/

#include <assert.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

/*----------------------------------------------------------------------------
 * Local headers
 *----------------------------------------------------------------------------*/

#include "cs_headers.h"

/*----------------------------------------------------------------------------*/

BEGIN_C_DECLS

/*----------------------------------------------------------------------------*/
/*!
 * \file cs_user_boundary_conditions-richards.c
 *
 * \brief User functions for input of calculation parameters.
 *
 * See \ref parameters for examples.
 */
/*----------------------------------------------------------------------------*/

/*=============================================================================
 * User function definitions
 *============================================================================*/

/*----------------------------------------------------------------------------*/
/*!
 * \brief User definition of boundary conditions
 *
 * \param[in, out]  domain   pointer to a cs_domain_t structure
 * \param[in, out]  bc_type  boundary face types
 *
 * The icodcl and rcodcl arrays are pre-initialized based on default
 * and GUI-defined definitions, and may be modified here.
 *
 * For a given variable field f, and a given face "face_id", these arrays
 * may be used as follows:
 *
 * - Boundary condition type code given at:
 *   f->bc_coeffs->icodcl[face_id]
 *
 * - Dirichlet value defined at:
 *   f->bc_coeffs->rcodcl1[face_id]
 *
 * - Interior exchange coefficient (infinite if no exchange) at:
 *   f->bc_coeffs->rcodcl2[face_id]
 *
 * - Flux density defined at:
 *   f->bc_coeffs->rcodcl3[face_id]
 *
 * For vector or tensor fields, these arrays are not interleaved,
 * so for a given face "face_id" and field component "comp_id", acess
 * is as follows (where n_b_faces is domain->mesh->n_b_faces):
 *
 *   f->bc_coeffs->rcodcl1[n_b_faces*comp_id + face_id]
 *   f->bc_coeffs->rcodcl2[n_b_faces*comp_id + face_id]
 *   f->bc_coeffs->rcodcl3[n_b_faces*comp_id + face_id]
 *
 * Only the icodcl code values from the first component are used in the case
 * of vector or tensor fields, so the icodcl values can be defined as for
 * a scalar.
 */
/*----------------------------------------------------------------------------*/

void
cs_user_boundary_conditions(cs_domain_t  *domain,
                            int           bc_type[])
{
  const cs_lnum_t n_b_faces = domain->mesh->n_b_faces;
  const cs_lnum_t *b_face_cells = domain->mesh->b_face_cells;

  const cs_real_3_t *restrict b_face_cog
    = (const cs_real_3_t *restrict)domain->mesh_quantities->b_face_cog;
  const cs_real_t *restrict b_face_surf
    = ( const cs_real_t *restrict)domain->mesh_quantities->b_face_surf;

  const cs_real_t t_cur = domain->time_step->t_cur;

  const int n_fields = cs_field_n_fields();
  const int keysca = cs_field_key_id("scalar_id");

  const cs_real_t *dt = CS_F_(dt)->val;

  const cs_zone_t *zn = NULL;

  /* Boundary mass flux */
  int iflmab = cs_field_get_key_int(CS_F_(p),
                                    cs_field_key_id("boundary_mass_flux_id"));
  const cs_real_t *b_mass_flux = cs_field_by_id(iflmab)->val;

  /* For groundwater flow module, the undefined type face iindef is always used.
   * BC for the flow part are on the pressure head H=h+z */

  /* Example 1: Dirichlet on hydraulic head and solutes */

  /*! [richards_bc_ex1] */
  zn = cs_boundary_zone_by_name("FACE1");

  for (cs_lnum_t e_idx = 0; e_idx < zn->n_elts; e_idx++) {

    const cs_lnum_t face_id = zn->elt_ids[e_idx];

    /* Undefined type face */
    bc_type[face_id] = CS_INDEF;

    /* Velocity is not solved but deduced from darcy law, so no BC is required.
     * However, no flux BC are imposed for safety. */

    CS_F_(vel)->bc_coeffs->icodcl[face_id] = 3;

    for (int ii = 0; ii< CS_F_(vel)->dim; ii++)
      CS_F_(vel)->bc_coeffs->rcodcl3[n_b_faces*ii + face_id] = 0;

    /* Dirichlet BC on hydraulic head (H = h + z) to impose a constant value */
    CS_F_(p)->bc_coeffs->icodcl[face_id] = 1;
    CS_F_(p)->bc_coeffs->rcodcl1[face_id] = 10.;

    /* Dirichlet BC on centration C (g/m^3) */
    for (int f_id = 0; f_id < n_fields; f_id++) {
      cs_field_t *f = cs_field_by_id(f_id);

      /* Here we only handle user scalars */
      if (cs_field_get_key_int(f, keysca) > 0) {
        f->bc_coeffs->icodcl[face_id] = 1;
        f->bc_coeffs->rcodcl1[face_id] = 1.0;
      }
    }
  }
  /*! [richards_bc_ex1] */

  /*! [richards_bc_ex2] */
  zn = cs_boundary_zone_by_name("FACE2");

  for (cs_lnum_t e_idx = 0; e_idx < zn->n_elts; e_idx++) {

    const cs_lnum_t face_id = zn->elt_ids[e_idx];

    /* Undefined type face */
    bc_type[face_id] = CS_INDEF;

    /* Dirichlet BC on hydraulic head (H = h + z) to impose a constant
       gradient over z axis here \f$ \grad H \cdot \vect{z} = -0.1 \f$*/
    CS_F_(p)->bc_coeffs->icodcl[face_id] = 1;
    CS_F_(p)->bc_coeffs->rcodcl1[face_id] = -1.*b_face_cog[face_id][2];
  }
  /*! [richards_bc_ex2] */

  /* Neumann on hydraulic head and solutes */

  /*! [richards_bc_ex3] */
  zn = cs_boundary_zone_by_name("FACE3");

  for (cs_lnum_t e_idx = 0; e_idx < zn->n_elts; e_idx++) {

    const cs_lnum_t face_id = zn->elt_ids[e_idx];

    /* Undefined type face */
    bc_type[face_id] = CS_INDEF;

    /* Neumann BC on hydraulic head (H = h + z) to impose a gradient along
       the surface normal
       Here \f$ \grad H \cdot \vect{n} = 0 \f$ */
    CS_F_(p)->bc_coeffs->icodcl[face_id] = 3;
    CS_F_(p)->bc_coeffs->rcodcl1[face_id] = 0.;

    /* Dirichlet BC on centration C (g/m^3) */
    for (int f_id = 0; f_id < n_fields; f_id++) {
      cs_field_t *f = cs_field_by_id(f_id);

      /* Here we only handle user scalars */
      if (cs_field_get_key_int(f, keysca) > 0) {

        /* Neumann BC on centration C for boundary surface with outward
         *  or null normal flow $V_out$:
         * It allows to impose a gradient along the surface normal for a
         * diffusive flux as the
         * convective flux is defined by the $C*V_out$.
         * Here \f$ \grad C \cdot \vect{n} = 0 \f$ */
        f->bc_coeffs->icodcl[face_id] = 3;
        f->bc_coeffs->rcodcl3[face_id] = 0.;

      }
    }
  }
  /*! [richards_bc_ex3] */

  /* Mixed (or Robin) BC to impose a total flux (diffusive and convective flux)*/

  /*! [richards_bc_ex4] */
  zn = cs_boundary_zone_by_name("FACE4");

  for (cs_lnum_t e_idx = 0; e_idx < zn->n_elts; e_idx++) {

    const cs_lnum_t face_id = zn->elt_ids[e_idx];
    const cs_lnum_t c_id = b_face_cells[face_id];

    /* Undefined type face */
    bc_type[face_id] = CS_INDEF;

    /* Dirichlet BC on hydraulic head (H = h + z) to impose a constant value */
    CS_F_(p)->bc_coeffs->icodcl[face_id] = 1;
    CS_F_(p)->bc_coeffs->rcodcl1[face_id] = 10.;

    /* For impose a radioactive activity R_act (Bq) at a surface S_act (m^2)
       during a period of time T_act, a mixed (or Robin) BC is used.
       We need two quantities:
         the velocity at an entrance is Vel = - mass_flux / surf
         or Vel = mass_flux / surf at an exit
         the reference concentration Cref = R_act / (S_act * dt * Vel) */

    const cs_real_t R_act = 0.5;
    const cs_real_t S_act = 50.0;
    const cs_real_t T_act = 10.0;

    /* The total flux is imposed from 0 to T_act */
    const cs_real_t Vel = -b_mass_flux[face_id]/b_face_surf[face_id];
    cs_real_t  Cref = R_act / (S_act * dt[c_id] * Vel);

    if (t_cur > T_act)
      Cref = 0.;

    for (int f_id = 0; f_id < n_fields; f_id++) {
      cs_field_t *f = cs_field_by_id(f_id);

      /* Here we only handle user scalar */
      if (cs_field_get_key_int(f, keysca) > 0) {
        f->bc_coeffs->icodcl[face_id] = 1;
        f->bc_coeffs->rcodcl1[face_id] = Cref;
        f->bc_coeffs->rcodcl2[face_id] = Vel;
      }
    }
  }
  /*! [richards_bc_ex4] */
}

/*----------------------------------------------------------------------------*/

END_C_DECLS
